# This is about Bankdata Cleansing with PySpark


In [1]:
#csv_df = spark.read.format('csv').option('header', True).option('inferSchema', True).option('escape', '"').load('abfss://65ca3b95-e765-425e-babe-29ac1e3c086a@onelake.dfs.fabric.microsoft.com/6d842f3b-7d08-4037-b1ec-7443755e8379/Files/accepted_2007_to_2018Q4.csv')

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 3, Finished, Available, Finished)

In [2]:
#display(csv_df.select("*").filter(col("id") == 61400928))

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 4, Finished, Available, Finished)

In [3]:
#csv_df.write.format('delta').saveAsTable('inv_bank_data')

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 5, Finished, Available, Finished)

In [4]:
from pyspark.sql.functions import *

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 6, Finished, Available, Finished)

In [5]:
df = spark.read.format('delta').load("abfss://65ca3b95-e765-425e-babe-29ac1e3c086a@onelake.dfs.fabric.microsoft.com/6d842f3b-7d08-4037-b1ec-7443755e8379/Tables/inv_bank_data",header=True,inferSchema=True)
display(df.head(2))

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 7, Finished, Available, Finished)

SynapseWidget(Synapse.DataFrame, b3613963-aea1-4192-a50d-451f07022d39)

In [6]:
display(df.select("*").filter(col("id") == 61400928)) #.show(5)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 8, Finished, Available, Finished)

SynapseWidget(Synapse.DataFrame, e7eee87f-cab4-4f4e-9e3f-d0a34021f341)

In [7]:
display(df.printSchema())

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 9, Finished, Available, Finished)

root
 |-- id: string (nullable = true)
 |-- member_id: string (nullable = true)
 |-- loan_amnt: double (nullable = true)
 |-- funded_amnt: double (nullable = true)
 |-- funded_amnt_inv: double (nullable = true)
 |-- term: string (nullable = true)
 |-- int_rate: double (nullable = true)
 |-- installment: double (nullable = true)
 |-- grade: string (nullable = true)
 |-- sub_grade: string (nullable = true)
 |-- emp_title: string (nullable = true)
 |-- emp_length: string (nullable = true)
 |-- home_ownership: string (nullable = true)
 |-- annual_inc: double (nullable = true)
 |-- verification_status: string (nullable = true)
 |-- issue_d: string (nullable = true)
 |-- loan_status: string (nullable = true)
 |-- pymnt_plan: string (nullable = true)
 |-- url: string (nullable = true)
 |-- desc: string (nullable = true)
 |-- purpose: string (nullable = true)
 |-- title: string (nullable = true)
 |-- zip_code: string (nullable = true)
 |-- addr_state: string (nullable = true)
 |-- dti: double 

In [8]:
list(df.columns)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 10, Finished, Available, Finished)

['id',
 'member_id',
 'loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'term',
 'int_rate',
 'installment',
 'grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'url',
 'desc',
 'purpose',
 'title',
 'zip_code',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'fico_range_low',
 'fico_range_high',
 'inq_last_6mths',
 'mths_since_last_delinq',
 'mths_since_last_record',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc',
 'initial_list_status',
 'out_prncp',
 'out_prncp_inv',
 'total_pymnt',
 'total_pymnt_inv',
 'total_rec_prncp',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'last_pymnt_d',
 'last_pymnt_amnt',
 'next_pymnt_d',
 'last_credit_pull_d',
 'last_fico_range_high',
 'last_fico_range_low',
 'collections_12_mths_ex_med',
 'mths_since_last_major_derog',
 'policy_code',
 'application_type',
 'annual_inc_joint',
 '

In [9]:
rows = df.count()
columns= len(df.columns)
print("Rows",rows,",","Columns",columns) #to find the rows and columns

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 11, Finished, Available, Finished)

Rows 2260701 , Columns 151


In [10]:
df.filter(col('id').isNull()).count() #to check the missing/null values

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 12, Finished, Available, Finished)

0

In [11]:
df.filter(col('member_id').isNull()).count() #to check the missing/null values

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 13, Finished, Available, Finished)

2260701

In [12]:
df.filter(col("loan_amnt").isNull()).count() #to check the missing/null values

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 14, Finished, Available, Finished)

33

In [13]:
df.filter(col("funded_amnt").isNull()).count() ##to check the missing/null values

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 15, Finished, Available, Finished)

33

In [14]:
def check_missing_values(data,cl):
    a = data.filter(col(cl).isNull()).count()
    b = data.count()
    c= (a/b) * 100
    return c                                         #to check the missing value percentages of each column passing

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 16, Finished, Available, Finished)

In [15]:
check_missing_values(df,'id')

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 17, Finished, Available, Finished)

0.0

In [16]:
check_missing_values(df,'member_id')

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 18, Finished, Available, Finished)

100.0

In [17]:
check_missing_values(df,"loan_amnt")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 19, Finished, Available, Finished)

0.0014597242182845054

In [18]:
check_missing_values(df,"funded_amnt")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 20, Finished, Available, Finished)

0.0014597242182845054

# to get all the columns missing values at single instance

In [19]:
#to get all the columns missing values at single instance
def check_miss_value_pctg(data,lst_cl):
    missing_values = {}
    for i in lst_cl:
        a = data.filter(col(i).isNull()).count()
        missing_values[i] = a
    return (missing_values)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 21, Finished, Available, Finished)

In [20]:
check_miss_value_pctg(df,df.columns)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 22, Finished, Available, Finished)

{'id': 0,
 'member_id': 2260701,
 'loan_amnt': 33,
 'funded_amnt': 33,
 'funded_amnt_inv': 33,
 'term': 33,
 'int_rate': 33,
 'installment': 33,
 'grade': 33,
 'sub_grade': 33,
 'emp_title': 167002,
 'emp_length': 146940,
 'home_ownership': 33,
 'annual_inc': 37,
 'verification_status': 33,
 'issue_d': 33,
 'loan_status': 33,
 'pymnt_plan': 33,
 'url': 33,
 'desc': 2134634,
 'purpose': 33,
 'title': 23358,
 'zip_code': 34,
 'addr_state': 33,
 'dti': 1744,
 'delinq_2yrs': 62,
 'earliest_cr_line': 62,
 'fico_range_low': 33,
 'fico_range_high': 33,
 'inq_last_6mths': 63,
 'mths_since_last_delinq': 1158535,
 'mths_since_last_record': 1901545,
 'open_acc': 62,
 'pub_rec': 62,
 'revol_bal': 33,
 'revol_util': 1835,
 'total_acc': 62,
 'initial_list_status': 33,
 'out_prncp': 33,
 'out_prncp_inv': 33,
 'total_pymnt': 33,
 'total_pymnt_inv': 33,
 'total_rec_prncp': 33,
 'total_rec_int': 33,
 'total_rec_late_fee': 33,
 'recoveries': 33,
 'collection_recovery_fee': 33,
 'last_pymnt_d': 2460,
 'la

# to get all the columns missing values percentage at single instance

In [21]:
#to get all the columns missing values percentage at single instance
def check_missing_value_pcntg(data,lst_cl):
    missing_values_less_than_75 = {}
    missing_values_greater_than_75 = {}
    b = data.count()
    for i in lst_cl:
        a = df.filter(col(i).isNull()).count()
        c = (a/b) * 100
        if c >= 75:
            missing_values_greater_than_75[i] = c
        else:
            missing_values_less_than_75[i] = c
    return ({"Missing_Value_above75:":missing_values_greater_than_75,"Missing_Value_below75:":missing_values_less_than_75})

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 23, Finished, Available, Finished)

In [22]:
check_missing_value_pcntg(df,df.columns) #checking the missing values percentage of each columns

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 24, Finished, Available, Finished)

{'Missing_Value_above75:': {'member_id': 100.0,
  'desc': 94.42354384768265,
  'mths_since_last_record': 84.11306935326698,
  'annual_inc_joint': 94.66050574578416,
  'dti_joint': 94.66068268205305,
  'verification_status_joint': 94.88079140054346,
  'mths_since_recent_bc_dlq': 77.01151103131285,
  'revol_bal_joint': 95.22183605881538,
  'sec_app_fico_range_low': 95.22179182474817,
  'sec_app_fico_range_high': 95.22179182474817,
  'sec_app_earliest_cr_line': 95.22179182474817,
  'sec_app_inq_last_6mths': 95.22179182474817,
  'sec_app_mort_acc': 95.22179182474817,
  'sec_app_open_acc': 95.22179182474817,
  'sec_app_revol_util': 95.30304980623266,
  'sec_app_open_act_il': 95.22179182474817,
  'sec_app_num_rev_accts': 95.22179182474817,
  'sec_app_chargeoff_within_12_mths': 95.22179182474817,
  'sec_app_collections_12_mths_ex_med': 95.22179182474817,
  'sec_app_mths_since_last_major_derog': 98.41013915595207,
  'hardship_type': 99.51709668815116,
  'hardship_reason': 99.51709668815116,
  

In [23]:
miss = check_missing_value_pcntg(df,df.columns) #assigning to df

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 25, Finished, Available, Finished)

In [24]:
column_names = list(miss['Missing_Value_above75:'].keys())
print("columnNames:",column_names)                             #extratcing columns that nulls more than 75%  

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 26, Finished, Available, Finished)

columnNames: ['member_id', 'desc', 'mths_since_last_record', 'annual_inc_joint', 'dti_joint', 'verification_status_joint', 'mths_since_recent_bc_dlq', 'revol_bal_joint', 'sec_app_fico_range_low', 'sec_app_fico_range_high', 'sec_app_earliest_cr_line', 'sec_app_inq_last_6mths', 'sec_app_mort_acc', 'sec_app_open_acc', 'sec_app_revol_util', 'sec_app_open_act_il', 'sec_app_num_rev_accts', 'sec_app_chargeoff_within_12_mths', 'sec_app_collections_12_mths_ex_med', 'sec_app_mths_since_last_major_derog', 'hardship_type', 'hardship_reason', 'hardship_status', 'deferral_term', 'hardship_amount', 'hardship_start_date', 'hardship_end_date', 'payment_plan_start_date', 'hardship_length', 'hardship_dpd', 'hardship_loan_status', 'orig_projected_additional_accrued_interest', 'hardship_payoff_balance_amount', 'hardship_last_payment_amount', 'debt_settlement_flag_date', 'settlement_status', 'settlement_date', 'settlement_amount', 'settlement_percentage', 'settlement_term']


# cols_nulls_morethan_75pct

In [25]:
cols_nulls_morethan_75pct = ['member_id', 'desc', 'mths_since_last_record', 'annual_inc_joint', 'dti_joint', 'verification_status_joint', 'mths_since_recent_bc_dlq', 'revol_bal_joint', 'sec_app_fico_range_low', 'sec_app_fico_range_high', 'sec_app_earliest_cr_line', 'sec_app_inq_last_6mths', 'sec_app_mort_acc', 'sec_app_open_acc', 'sec_app_revol_util', 'sec_app_open_act_il', 'sec_app_num_rev_accts', 'sec_app_chargeoff_within_12_mths', 'sec_app_collections_12_mths_ex_med', 'sec_app_mths_since_last_major_derog', 'hardship_type', 'hardship_reason', 'hardship_status', 'deferral_term', 'hardship_amount', 'hardship_start_date', 'hardship_end_date', 'payment_plan_start_date', 'hardship_length', 'hardship_dpd', 'hardship_loan_status', 'orig_projected_additional_accrued_interest', 'hardship_payoff_balance_amount', 'hardship_last_payment_amount', 'debt_settlement_flag_date', 'settlement_status', 'settlement_date', 'settlement_amount', 'settlement_percentage', 'settlement_term']

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 27, Finished, Available, Finished)

# To chec duplicates 

In [26]:
def check_dups(data,cl):
    a = data.select(cl).distinct().count()
    b = data.count()
    if a == b:
        print("No Duplicates")
    else:
        c = b - a
        print("There are", c, "Duplicates")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 28, Finished, Available, Finished)

In [27]:
check_dups(df,"id")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 29, Finished, Available, Finished)

No Duplicates


In [28]:
check_dups(df,"num_bc_tl")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 30, Finished, Available, Finished)

There are 2260624 Duplicates


# To check all column Duplicates at single instance

In [29]:
def duplicate_check_allclms(data):
    for cl in df.columns:                            #writing fro loop
        a = data.select(cl).distinct().count()     #define variable
        b = data.count()                             #define variable
        if a == b:
            print(f"No Duplicates in column: {cl}")
        else:
            c = b-a
            print(f"There are {c} duplicates in column: {cl}")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 31, Finished, Available, Finished)

In [30]:
duplicate_check_allclms(df)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 32, Finished, Available, Finished)

No Duplicates in column: id
There are 2260700 duplicates in column: member_id
There are 2259128 duplicates in column: loan_amnt
There are 2259128 duplicates in column: funded_amnt
There are 2250643 duplicates in column: funded_amnt_inv
There are 2260698 duplicates in column: term
There are 2260027 duplicates in column: int_rate
There are 2167399 duplicates in column: installment
There are 2260693 duplicates in column: grade
There are 2260665 duplicates in column: sub_grade
There are 1748006 duplicates in column: emp_title
There are 2260689 duplicates in column: emp_length
There are 2260694 duplicates in column: home_ownership
There are 2171332 duplicates in column: annual_inc
There are 2260697 duplicates in column: verification_status
There are 2260561 duplicates in column: issue_d
There are 2260691 duplicates in column: loan_status
There are 2260698 duplicates in column: pymnt_plan
There are 32 duplicates in column: url
There are 2136199 duplicates in column: desc
There are 2260686 du

In [31]:
df.select("id").distinct().orderBy(asc('id')).show(5)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 33, Finished, Available, Finished)

+---------+
|       id|
+---------+
|  1000007|
|100001133|
|100001137|
|100001142|
|100001158|
+---------+
only showing top 5 rows



In [32]:
df.select('id').distinct().orderBy(desc('id')).show(40,False)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 34, Finished, Available, Finished)

+------------------------------------------------+
|id                                              |
+------------------------------------------------+
|Total amount funded in policy code 2: 873652739 |
|Total amount funded in policy code 2: 823319310 |
|Total amount funded in policy code 2: 820109297 |
|Total amount funded in policy code 2: 81866225  |
|Total amount funded in policy code 2: 737901574 |
|Total amount funded in policy code 2: 662815446 |
|Total amount funded in policy code 2: 651669342 |
|Total amount funded in policy code 2: 620899600 |
|Total amount funded in policy code 2: 608903141 |
|Total amount funded in policy code 2: 567447023 |
|Total amount funded in policy code 2: 564202131 |
|Total amount funded in policy code 2: 521953170 |
|Total amount funded in policy code 2: 520780182 |
|Total amount funded in policy code 2: 511988838 |
|Total amount funded in policy code 2: 1944088810|
|Total amount funded in policy code 2: 0         |
|Total amount funded in policy 

In [33]:
df.select('id').filter(col('id').like("%Total amount funded in policy code%")).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 35, Finished, Available, Finished)

32

In [34]:
display(df.filter(col('id').like("%Total amount funded in policy code%")))

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 36, Finished, Available, Finished)

SynapseWidget(Synapse.DataFrame, 0a049002-f7bb-40a0-b499-ae502b5c8977)

In [35]:
df.select("*").filter(~col('id').rlike("^[A-Za-z]")).count() #to filter the null rows from ID column having string values

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 37, Finished, Available, Finished)

2260668

In [36]:
df = df.filter(~col('id').rlike("^[A-Za-z]"))

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 38, Finished, Available, Finished)

In [37]:
df.count() #33 rows removed having nulls  2260701 - 2260668

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 39, Finished, Available, Finished)

2260668

In [38]:
df.select("term").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 40, Finished, Available, Finished)

+----------+
|      term|
+----------+
| 36 months|
| 60 months|
+----------+



In [39]:
df = df.withColumn("term_in_months", regexp_replace(col("term"), " months",""))

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 41, Finished, Available, Finished)

In [40]:
df.select("term_in_months").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 42, Finished, Available, Finished)

+--------------+
|term_in_months|
+--------------+
|            60|
|            36|
+--------------+



# Columns to Drop
Term, emp_length,url,mths_since_last_major_derog

In [41]:
#check the column have string datatype
for cl, dtype in df.dtypes:
    if dtype == 'string':
        print(f"Column '{cl}' has string values")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 43, Finished, Available, Finished)

Column 'id' has string values
Column 'member_id' has string values
Column 'term' has string values
Column 'grade' has string values
Column 'sub_grade' has string values
Column 'emp_title' has string values
Column 'emp_length' has string values
Column 'home_ownership' has string values
Column 'verification_status' has string values
Column 'issue_d' has string values
Column 'loan_status' has string values
Column 'pymnt_plan' has string values
Column 'url' has string values
Column 'desc' has string values
Column 'purpose' has string values
Column 'title' has string values
Column 'zip_code' has string values
Column 'addr_state' has string values
Column 'earliest_cr_line' has string values
Column 'initial_list_status' has string values
Column 'last_pymnt_d' has string values
Column 'next_pymnt_d' has string values
Column 'last_credit_pull_d' has string values
Column 'application_type' has string values
Column 'verification_status_joint' has string values
Column 'sec_app_earliest_cr_line' ha

In [42]:
display(df.printSchema())

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 44, Finished, Available, Finished)

root
 |-- id: string (nullable = true)
 |-- member_id: string (nullable = true)
 |-- loan_amnt: double (nullable = true)
 |-- funded_amnt: double (nullable = true)
 |-- funded_amnt_inv: double (nullable = true)
 |-- term: string (nullable = true)
 |-- int_rate: double (nullable = true)
 |-- installment: double (nullable = true)
 |-- grade: string (nullable = true)
 |-- sub_grade: string (nullable = true)
 |-- emp_title: string (nullable = true)
 |-- emp_length: string (nullable = true)
 |-- home_ownership: string (nullable = true)
 |-- annual_inc: double (nullable = true)
 |-- verification_status: string (nullable = true)
 |-- issue_d: string (nullable = true)
 |-- loan_status: string (nullable = true)
 |-- pymnt_plan: string (nullable = true)
 |-- url: string (nullable = true)
 |-- desc: string (nullable = true)
 |-- purpose: string (nullable = true)
 |-- title: string (nullable = true)
 |-- zip_code: string (nullable = true)
 |-- addr_state: string (nullable = true)
 |-- dti: double 

In [43]:
df.select('grade').distinct().show() #getting distinct values in grade column

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 45, Finished, Available, Finished)

+-----+
|grade|
+-----+
|    F|
|    E|
|    B|
|    D|
|    C|
|    A|
|    G|
+-----+



In [44]:
#to confirm if grade have only 1 character
df.select("grade").filter(length(col("grade")) >1).show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 46, Finished, Available, Finished)

+-----+
|grade|
+-----+
+-----+



In [45]:
df.select("sub_grade").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 47, Finished, Available, Finished)

+---------+
|sub_grade|
+---------+
|       D5|
|       F2|
|       B4|
|       A2|
|       E4|
|       B2|
|       C3|
|       D1|
|       C4|
|       F1|
|       D3|
|       F5|
|       G2|
|       B1|
|       B3|
|       E5|
|       C5|
|       G3|
|       A4|
|       F4|
+---------+
only showing top 20 rows



In [46]:
df.select("sub_grade").filter(length(col("sub_grade"))>2).show() ##to confirm if sub_grade have only 2 character

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 48, Finished, Available, Finished)

+---------+
|sub_grade|
+---------+
+---------+



In [47]:
df.select("sub_grade").filter(length(col("sub_grade"))<2).show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 49, Finished, Available, Finished)

+---------+
|sub_grade|
+---------+
+---------+



In [48]:
df.select("emp_title").filter(col('emp_title').isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 50, Finished, Available, Finished)

166969

In [49]:
df.select("emp_title").distinct().show(10,False)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 51, Finished, Available, Finished)

+-------------------------+
|emp_title                |
+-------------------------+
|Systems Administrator II |
|Physician                |
|CSR                      |
|Nutrition                |
|tool room attendant      |
|office admin             |
|Front End Web Developer  |
|machinist                |
|SUPERINTENDENT           |
|Implementation Consultant|
+-------------------------+
only showing top 10 rows



In [50]:
df.groupBy("emp_title").count().orderBy(desc("count")).show(20)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 52, Finished, Available, Finished)

+------------------+------+
|         emp_title| count|
+------------------+------+
|              NULL|166969|
|           Teacher| 38824|
|           Manager| 34298|
|             Owner| 21977|
|  Registered Nurse| 15867|
|            Driver| 14753|
|                RN| 14737|
|        Supervisor| 14297|
|             Sales| 13050|
|   Project Manager| 10971|
|    Office Manager|  9772|
|   General Manager|  9251|
|          Director|  8934|
|             owner|  8507|
|         President|  7660|
|          Engineer|  7304|
|           manager|  7060|
|           teacher|  6692|
|Operations Manager|  6128|
|    Vice President|  5874|
+------------------+------+
only showing top 20 rows



In [51]:
df = df.fillna("Others", subset = ["emp_title"]) #fill nulls with some other values in a columns by using the subset

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 53, Finished, Available, Finished)

In [52]:
df.groupBy("emp_title").count().orderBy(desc("count")).show(20)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 54, Finished, Available, Finished)

+------------------+------+
|         emp_title| count|
+------------------+------+
|            Others|166969|
|           Teacher| 38824|
|           Manager| 34298|
|             Owner| 21977|
|  Registered Nurse| 15867|
|            Driver| 14753|
|                RN| 14737|
|        Supervisor| 14297|
|             Sales| 13050|
|   Project Manager| 10971|
|    Office Manager|  9772|
|   General Manager|  9251|
|          Director|  8934|
|             owner|  8507|
|         President|  7660|
|          Engineer|  7304|
|           manager|  7060|
|           teacher|  6692|
|Operations Manager|  6128|
|    Vice President|  5874|
+------------------+------+
only showing top 20 rows



In [53]:
display(df.select("*").filter(col("emp_title").isNull())) #.show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 55, Finished, Available, Finished)

SynapseWidget(Synapse.DataFrame, 83c9405f-e89d-4e78-839f-0ad43e02aa47)

In [54]:
df = df.withColumn("emp_title",initcap(col("emp_title"))) #to capitalize the string vlaues in a column

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 56, Finished, Available, Finished)

In [55]:
df.groupBy("emp_title").count().orderBy(desc("count")).show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 57, Finished, Available, Finished)

+----------------+------+
|       emp_title| count|
+----------------+------+
|          Others|166970|
|         Teacher| 46125|
|         Manager| 42822|
|           Owner| 31740|
|Registered Nurse| 21407|
|          Driver| 20786|
|      Supervisor| 18560|
|           Sales| 17647|
|              Rn| 16672|
|  Office Manager| 13163|
+----------------+------+
only showing top 10 rows



In [56]:
df.select('emp_title').distinct().count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 58, Finished, Available, Finished)

438350

In [57]:
df.select("emp_length").distinct().show(20,False)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 59, Finished, Available, Finished)

+----------+
|emp_length|
+----------+
|9 years   |
|5 years   |
|1 year    |
|2 years   |
|7 years   |
|8 years   |
|4 years   |
|6 years   |
|3 years   |
|10+ years |
|< 1 year  |
|NULL      |
+----------+



In [58]:
#display(df.select("*").filter(col("emp_length") == ' reactors"')) #.show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 60, Finished, Available, Finished)

In [59]:
df.select("home_ownership").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 61, Finished, Available, Finished)

+--------------+
|home_ownership|
+--------------+
|           OWN|
|          RENT|
|      MORTGAGE|
|           ANY|
|         OTHER|
|          NONE|
+--------------+



In [60]:
df = df.withColumn("emp_length_years",regexp_extract(col("emp_length"), "\\d+",0)) #to extract the digits from emp_lengthcolumn

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 62, Finished, Available, Finished)

In [61]:
df.select("emp_length","emp_length_years").distinct().show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 63, Finished, Available, Finished)

+----------+----------------+
|emp_length|emp_length_years|
+----------+----------------+
|   6 years|               6|
|   3 years|               3|
| 10+ years|              10|
|   7 years|               7|
|  < 1 year|               1|
|   4 years|               4|
|   8 years|               8|
|   2 years|               2|
|   9 years|               9|
|    1 year|               1|
+----------+----------------+
only showing top 10 rows



In [62]:
df.select("*").filter(col("emp_length_years").isNull()).count() # teo get the null values count in emp_length_years column

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 64, Finished, Available, Finished)

146907

In [63]:
display(df.filter(col("emp_length_years").isNull()).head(5)) #.show(20)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 65, Finished, Available, Finished)

SynapseWidget(Synapse.DataFrame, d7a6d30f-b003-4a71-9621-798da8e10c7f)

In [64]:
df = df.fillna("NA", subset = ["emp_length_years"]) #replaced the null values with NA

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 66, Finished, Available, Finished)

In [65]:
df.select("*").filter(col("emp_length_years").isNull()).count() # now there is no null values

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 67, Finished, Available, Finished)

0

In [66]:
df.groupBy(col("emp_length_years")).count().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 68, Finished, Available, Finished)

+----------------+------+
|emp_length_years| count|
+----------------+------+
|               7| 92695|
|               3|180753|
|               8| 91914|
|              NA|146907|
|               5|139698|
|               6|102628|
|               9| 79395|
|               1|338391|
|              10|748005|
|               4|136605|
|               2|203677|
+----------------+------+



In [67]:
df.select("home_ownership").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 69, Finished, Available, Finished)

+--------------+
|home_ownership|
+--------------+
|           OWN|
|          RENT|
|      MORTGAGE|
|           ANY|
|         OTHER|
|          NONE|
+--------------+



In [68]:
df.select("verification_status").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 70, Finished, Available, Finished)

+-------------------+
|verification_status|
+-------------------+
|           Verified|
|    Source Verified|
|       Not Verified|
+-------------------+



In [69]:
df.select("loan_status").distinct().show(10,False)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 71, Finished, Available, Finished)

+---------------------------------------------------+
|loan_status                                        |
+---------------------------------------------------+
|Fully Paid                                         |
|Default                                            |
|In Grace Period                                    |
|Charged Off                                        |
|Late (31-120 days)                                 |
|Current                                            |
|Late (16-30 days)                                  |
|Does not meet the credit policy. Status:Fully Paid |
|Does not meet the credit policy. Status:Charged Off|
+---------------------------------------------------+



In [70]:
df.select("title").distinct().count() #show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 72, Finished, Available, Finished)

63156

In [71]:
df.select("title").distinct().orderBy(desc(col("title"))).show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 73, Finished, Available, Finished)

+--------------------+
|               title|
+--------------------+
|MY FIRST CA...|
|        ~Summer Fun~|
|~Life Reorganizat...|
|               zxcvb|
|        zonball Loan|
|              zipcar|
|          zeusamoose|
|            zerodebt|
|       zero interest|
|           zero dept|
|           zero debt|
|zero credit card ...|
|        zero balance|
|                zero|
|          zandercade|
|                zack|
|your rate is bett...|
|     your helping me|
|   youngest daughter|
|young woman with ...|
+--------------------+
only showing top 20 rows



In [72]:
df.select(col('loan_amnt').isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 74, Finished, Available, Finished)

2260668

In [73]:
df=df.fillna(0, subset = "loan_amnt")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 75, Finished, Available, Finished)

In [74]:
df.select("loan_amnt").filter(col("loan_amnt").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 76, Finished, Available, Finished)

0

In [75]:
df.select("loan_amnt").distinct().orderBy(desc(col("loan_amnt"))).show(50)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 77, Finished, Available, Finished)

+---------+
|loan_amnt|
+---------+
|  40000.0|
|  39975.0|
|  39950.0|
|  39925.0|
|  39900.0|
|  39875.0|
|  39850.0|
|  39825.0|
|  39800.0|
|  39775.0|
|  39750.0|
|  39725.0|
|  39700.0|
|  39675.0|
|  39650.0|
|  39625.0|
|  39600.0|
|  39575.0|
|  39550.0|
|  39525.0|
|  39500.0|
|  39475.0|
|  39450.0|
|  39425.0|
|  39400.0|
|  39375.0|
|  39350.0|
|  39325.0|
|  39300.0|
|  39275.0|
|  39250.0|
|  39225.0|
|  39200.0|
|  39175.0|
|  39150.0|
|  39125.0|
|  39100.0|
|  39075.0|
|  39050.0|
|  39025.0|
|  39000.0|
|  38975.0|
|  38950.0|
|  38925.0|
|  38900.0|
|  38875.0|
|  38850.0|
|  38825.0|
|  38800.0|
|  38775.0|
+---------+
only showing top 50 rows



In [76]:
df.groupBy("loan_amnt").count().orderBy(desc(col("loan_amnt"))).show(5)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 78, Finished, Available, Finished)

+---------+-----+
|loan_amnt|count|
+---------+-----+
|  40000.0|33368|
|  39975.0|   11|
|  39950.0|   10|
|  39925.0|   14|
|  39900.0|   24|
+---------+-----+
only showing top 5 rows



In [77]:
df = df.fillna(0, subset = ["funded_amnt","funded_amnt_inv","int_rate","installment","annual_inc"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 79, Finished, Available, Finished)

In [78]:
df.select("verification_status").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 80, Finished, Available, Finished)

+-------------------+
|verification_status|
+-------------------+
|           Verified|
|    Source Verified|
|       Not Verified|
+-------------------+



In [79]:
df.select("issue_d").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 81, Finished, Available, Finished)

+--------+
| issue_d|
+--------+
|Oct-2016|
|Sep-2017|
|May-2015|
|Dec-2014|
|Mar-2018|
|Sep-2018|
|Jul-2015|
|Feb-2014|
|Sep-2015|
|Jan-2016|
|Nov-2017|
|Jan-2014|
|Jul-2018|
|Oct-2015|
|May-2016|
|Jan-2018|
|May-2014|
|Aug-2018|
|Apr-2014|
|Apr-2016|
+--------+
only showing top 20 rows



In [80]:
df.select("loan_status").distinct().show(10,False)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 82, Finished, Available, Finished)

+---------------------------------------------------+
|loan_status                                        |
+---------------------------------------------------+
|Fully Paid                                         |
|Default                                            |
|In Grace Period                                    |
|Charged Off                                        |
|Late (31-120 days)                                 |
|Current                                            |
|Late (16-30 days)                                  |
|Does not meet the credit policy. Status:Fully Paid |
|Does not meet the credit policy. Status:Charged Off|
+---------------------------------------------------+



In [81]:
df.select("pymnt_plan").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 83, Finished, Available, Finished)

+----------+
|pymnt_plan|
+----------+
|         n|
|         y|
+----------+



In [82]:
df.select("purpose").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 84, Finished, Available, Finished)

+------------------+
|           purpose|
+------------------+
|           wedding|
|             other|
|    small_business|
|debt_consolidation|
|       credit_card|
|            moving|
|          vacation|
|  renewable_energy|
|             house|
|               car|
|    major_purchase|
|           medical|
|  home_improvement|
|       educational|
+------------------+



In [83]:
df.groupby("purpose").count().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 85, Finished, Available, Finished)

+------------------+-------+
|           purpose|  count|
+------------------+-------+
|           wedding|   2355|
|             other| 139440|
|    small_business|  24689|
|debt_consolidation|1277877|
|       credit_card| 516971|
|            moving|  15403|
|          vacation|  15525|
|  renewable_energy|   1445|
|             house|  14136|
|               car|  24013|
|    major_purchase|  50445|
|           medical|  27488|
|  home_improvement| 150457|
|       educational|    424|
+------------------+-------+



In [84]:
df.groupBy("title").count().show(50)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 86, Finished, Available, Finished)

+--------------------+------+
|               title| count|
+--------------------+------+
|       Payoff Card's|     1|
|Davey's consolida...|     1|
|Credit Consolidation|   435|
|              Payoff|   666|
|    Home Improvement|  1773|
|                loan|   453|
|              floors|     2|
|             My Loan|   566|
|            Personal|  1185|
|  Credit Card Payoff|  1386|
|my debt consolida...|    17|
|  Debt Consolidation| 15763|
|                NULL| 23325|
|                Bill|     7|
|             My loan|   246|
|              School|    21|
|      Major purchase| 44840|
|credit card refin...|   492|
|          second try|     3|
|Credit Card Conso...|  2360|
| Bill consolidation |     9|
|           Debt Help|    58|
|    Debt elimination|    10|
|       Personal Loan|  2133|
|       credit cards |    18|
|Out of Debt With ...|     1|
|              debt 1|     3|
|                refi|   107|
|         consolidate|   763|
|               Other|127714|
|       mr

In [85]:
df.select("zip_code").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 87, Finished, Available, Finished)

+--------+
|zip_code|
+--------+
|   471xx|
|   418xx|
|   957xx|
|   223xx|
|   230xx|
|   143xx|
|   751xx|
|   154xx|
|   371xx|
|   591xx|
|   183xx|
|   831xx|
|   756xx|
|   287xx|
|   179xx|
|   535xx|
|   216xx|
|   895xx|
|   625xx|
|   387xx|
+--------+
only showing top 20 rows



In [86]:
df.select("zip_code").filter(length(col("zip_code")) > 5).show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 88, Finished, Available, Finished)

+--------+
|zip_code|
+--------+
+--------+



In [87]:
df.select("zip_code").filter(length(col("zip_code")) <5).show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 89, Finished, Available, Finished)

+--------+
|zip_code|
+--------+
+--------+



In [88]:
df.select("zip_code").filter(col("zip_code").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 90, Finished, Available, Finished)

1

In [89]:
df = df.fillna(0, subset = "zip_code")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 91, Finished, Available, Finished)

In [90]:
df.select("addr_state").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 92, Finished, Available, Finished)

+----------+
|addr_state|
+----------+
|        SC|
|        AZ|
|        LA|
|        MN|
|        NJ|
|        DC|
|        OR|
|        VA|
|        RI|
|        KY|
|        WY|
|        NH|
|        MI|
|        NV|
|        WI|
|        ID|
|        CA|
|        NE|
|        CT|
|        MT|
+----------+
only showing top 20 rows



In [91]:
df.select("addr_state").filter(length(col("addr_state")) > 2).show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 93, Finished, Available, Finished)

+----------+
|addr_state|
+----------+
+----------+



In [92]:
df.select("addr_state").filter(length(col("addr_state"))<2).show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 94, Finished, Available, Finished)

+----------+
|addr_state|
+----------+
+----------+



In [93]:
df.select("addr_state").filter(col("addr_state").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 95, Finished, Available, Finished)

0

In [94]:
df.select("dti").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 96, Finished, Available, Finished)

+-----+
|  dti|
+-----+
|19.98|
| 9.13|
|30.49|
| 14.9|
|17.52|
|17.56|
| 13.4|
| 2.86|
|23.04|
|37.81|
| 8.51|
| 3.26|
|17.95|
|35.17|
| 15.5|
|26.72|
| 26.7|
|12.32|
|41.89|
|38.61|
+-----+
only showing top 20 rows



In [95]:
df.select("dti").filter(col("dti").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 97, Finished, Available, Finished)

1711

In [96]:
df=df.fillna(0, subset = "dti")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 98, Finished, Available, Finished)

In [97]:
df.select("dti").filter(col("dti").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 99, Finished, Available, Finished)

0

In [98]:
df.select("delinq_2yrs").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 100, Finished, Available, Finished)

+-----------+
|delinq_2yrs|
+-----------+
|        8.0|
|        0.0|
|        7.0|
|       29.0|
|       35.0|
|       18.0|
|        1.0|
|        4.0|
|       11.0|
|       58.0|
|       21.0|
|       14.0|
|       22.0|
|        3.0|
|       19.0|
|       28.0|
|        2.0|
|       17.0|
|       27.0|
|       10.0|
+-----------+
only showing top 20 rows



In [99]:
df.select("delinq_2yrs").filter(col("delinq_2yrs").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 101, Finished, Available, Finished)

29

In [100]:
df=df.fillna(0, subset = "delinq_2yrs")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 102, Finished, Available, Finished)

In [101]:
df.select("delinq_2yrs").filter(col("delinq_2yrs").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 103, Finished, Available, Finished)

0

In [102]:
df.select("earliest_cr_line").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 104, Finished, Available, Finished)

+----------------+
|earliest_cr_line|
+----------------+
|        Jan-1999|
|        Jul-1996|
|        Nov-1978|
|        May-1977|
|        Mar-1999|
|        Jul-1989|
|        Mar-1960|
|        Sep-1987|
|        Jun-1979|
|        May-1973|
|        Apr-1988|
|        Oct-1975|
|        May-1993|
|        Sep-1998|
|        Jun-1985|
|        Feb-1961|
|        Jan-1953|
|        Dec-1981|
|        Jun-1989|
|        Oct-1965|
+----------------+
only showing top 20 rows



In [103]:
df.select("earliest_cr_line").filter(col("earliest_cr_line").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 105, Finished, Available, Finished)

29

In [104]:
df=df.fillna("NA",subset = "earliest_cr_line")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 106, Finished, Available, Finished)

In [105]:
df.select("earliest_cr_line").filter(col("earliest_cr_line").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 107, Finished, Available, Finished)

0

In [106]:
df.select("fico_range_low").filter(col("fico_range_low").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 108, Finished, Available, Finished)

0

In [107]:
df.select("fico_range_high").filter(col("fico_range_high").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 109, Finished, Available, Finished)

0

In [108]:
df.select("inq_last_6mths").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 110, Finished, Available, Finished)

+--------------+
|inq_last_6mths|
+--------------+
|           0.0|
|           1.0|
|          NULL|
|           4.0|
|           3.0|
|           2.0|
|           6.0|
|           5.0|
|           8.0|
|           7.0|
|          18.0|
|          25.0|
|          31.0|
|          11.0|
|          14.0|
|          19.0|
|          28.0|
|          17.0|
|          27.0|
|          10.0|
+--------------+
only showing top 20 rows



In [109]:
df.select("inq_last_6mths").filter(col("inq_last_6mths").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 111, Finished, Available, Finished)

30

In [110]:
df = df.fillna(0, subset = "inq_last_6mths")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 112, Finished, Available, Finished)

In [111]:
df.select("inq_last_6mths").filter(col("inq_last_6mths").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 113, Finished, Available, Finished)

0

In [112]:
df.select("mths_since_last_delinq").filter(col("mths_since_last_delinq").isNull()).count() #

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 114, Finished, Available, Finished)

1158502

In [113]:
df.select("open_acc").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 115, Finished, Available, Finished)

+--------+
|open_acc|
+--------+
|     8.0|
|     0.0|
|     7.0|
|    49.0|
|    29.0|
|    64.0|
|    47.0|
|    42.0|
|    44.0|
|    35.0|
|    62.0|
|    18.0|
|     1.0|
|    39.0|
|    34.0|
|    37.0|
|    25.0|
|    36.0|
|    41.0|
|     4.0|
+--------+
only showing top 20 rows



In [114]:
df.select("open_acc").filter(col("open_acc").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 116, Finished, Available, Finished)

29

In [115]:
df.select("pub_rec").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 117, Finished, Available, Finished)

+-------+
|pub_rec|
+-------+
|    8.0|
|    0.0|
|    7.0|
|   18.0|
|    1.0|
|   37.0|
|    4.0|
|   23.0|
|   11.0|
|   21.0|
|   14.0|
|   63.0|
|   22.0|
|    3.0|
|   19.0|
|    2.0|
|   17.0|
|   10.0|
|   40.0|
|   13.0|
+-------+
only showing top 20 rows



In [116]:
df.select("pub_rec").filter(col("pub_rec").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 118, Finished, Available, Finished)

29

In [117]:
df.select("revol_bal").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 119, Finished, Available, Finished)

+---------+
|revol_bal|
+---------+
|  64193.0|
|   2862.0|
|    299.0|
|   5360.0|
|   4142.0|
|  28980.0|
|  17072.0|
|  39763.0|
|  14473.0|
|  32046.0|
|  20689.0|
|  21825.0|
|   6454.0|
|  16822.0|
|   8649.0|
|   7115.0|
|  10625.0|
|  11757.0|
|  11772.0|
|   1761.0|
+---------+
only showing top 20 rows



In [118]:
df.select("revol_bal").filter(col("revol_bal").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 120, Finished, Available, Finished)

0

In [119]:
df = df.fillna(0, subset = ["mths_since_last_delinq","pub_rec","open_acc"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 121, Finished, Available, Finished)

In [120]:
df.select("mths_since_last_delinq").filter(col("mths_since_last_delinq").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 122, Finished, Available, Finished)

0

In [121]:
df.select("pub_rec").filter(col("pub_rec").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 123, Finished, Available, Finished)

0

In [122]:
df.select("open_acc").filter(col("open_acc").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 124, Finished, Available, Finished)

0

In [123]:
df.select(["revol_util","initial_list_status","total_acc","out_prncp","out_prncp_inv"]).show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 125, Finished, Available, Finished)

+----------+-------------------+---------+---------+-------------+
|revol_util|initial_list_status|total_acc|out_prncp|out_prncp_inv|
+----------+-------------------+---------+---------+-------------+
|      72.0|                  w|     18.0|      0.0|          0.0|
|      58.0|                  f|     16.0|      0.0|          0.0|
|      84.0|                  w|     35.0|      0.0|          0.0|
|      59.0|                  w|     40.0|      0.0|          0.0|
|      87.0|                  w|     25.0|      0.0|          0.0|
|      89.0|                  w|     10.0|      0.0|          0.0|
|      95.0|                  w|     24.0|      0.0|          0.0|
|      94.0|                  w|     37.0|      0.0|          0.0|
|      95.0|                  w|     13.0|      0.0|          0.0|
|      57.0|                  w|     20.0|      0.0|          0.0|
+----------+-------------------+---------+---------+-------------+
only showing top 10 rows



In [124]:
df = df.fillna(0, subset = ["revol_util","total_acc","out_prncp","out_prncp_inv"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 126, Finished, Available, Finished)

In [125]:
df.select("initial_list_status").distinct().show(20)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 127, Finished, Available, Finished)

+-------------------+
|initial_list_status|
+-------------------+
|                  f|
|                  w|
+-------------------+



In [126]:
df.select("initial_list_status").filter(col("initial_list_status").isNull()).count()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 128, Finished, Available, Finished)

0

In [127]:
df.select(["total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee"]).show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 129, Finished, Available, Finished)

+-----------+---------------+---------------+-------------+------------------+
|total_pymnt|total_pymnt_inv|total_rec_prncp|total_rec_int|total_rec_late_fee|
+-----------+---------------+---------------+-------------+------------------+
|   14654.32|        14629.9|        6591.69|      5076.65|               0.0|
|    7021.05|        7021.05|        4370.75|      2022.29|               0.0|
|   10490.93|       10490.93|         4707.5|      3176.95|               0.0|
|   18783.18|       18783.18|        7127.22|      6564.21|               0.0|
|     4393.9|         4393.9|        2267.74|      1330.64|               0.0|
|     4422.4|         4422.4|        1336.56|       907.46|               0.0|
|   23745.99|       23745.99|       15765.81|      5535.75|               0.0|
|    4772.89|        4772.89|         794.07|      2028.97|               0.0|
|    2617.01|        2617.01|        1682.61|       638.18|               0.0|
|    8165.46|        8165.46|        3837.23|      1

In [128]:
df = df.fillna(0,subset=["total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 130, Finished, Available, Finished)

In [130]:
df.select(["recoveries","collection_recovery_fee","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt"]).show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 132, Finished, Available, Finished)

+----------+-----------------------+-----------------------+------------+---------------+
|recoveries|collection_recovery_fee|collection_recovery_fee|last_pymnt_d|last_pymnt_amnt|
+----------+-----------------------+-----------------------+------------+---------------+
|   2985.98|               537.4764|               537.4764|    Feb-2017|         687.68|
|    628.01|               113.0418|               113.0418|    Mar-2017|         356.48|
|   2606.48|               469.1664|               469.1664|    Aug-2016|         660.24|
|   5091.75|                916.515|                916.515|    Mar-2017|         723.34|
|    795.52|               143.1936|               143.1936|    May-2016|         360.83|
|   2178.38|               392.1084|               392.1084|    Dec-2015|         551.64|
|   2444.43|               439.9974|               439.9974|    Jan-2017|        1184.86|
|   1949.85|                350.973|                350.973|    Oct-2015|         972.74|
|    296.2

In [131]:
df = df.fillna(0, subset = ["recoveries","collection_recovery_fee","collection_recovery_fee","last_pymnt_amnt"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 133, Finished, Available, Finished)

In [132]:
df.select(["next_pymnt_d","last_credit_pull_d"]).show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 134, Finished, Available, Finished)

+------------+------------------+
|next_pymnt_d|last_credit_pull_d|
+------------+------------------+
|        NULL|          Aug-2017|
|        NULL|          Jul-2018|
|        NULL|          Feb-2017|
|        NULL|          Aug-2017|
|        NULL|          Dec-2016|
|        NULL|          Oct-2016|
|        NULL|          Jul-2017|
|        NULL|          Oct-2017|
|        NULL|          Jun-2017|
|        NULL|          Oct-2016|
+------------+------------------+
only showing top 10 rows



In [133]:
df =df.fillna("NA", subset = ["last_pymnt_d","next_pymnt_d","last_credit_pull_d"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 135, Finished, Available, Finished)

In [134]:
df.select(["last_fico_range_high","last_fico_range_low","collections_12_mths_ex_med","mths_since_last_major_derog"]).show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 136, Finished, Available, Finished)

+--------------------+-------------------+--------------------------+---------------------------+
|last_fico_range_high|last_fico_range_low|collections_12_mths_ex_med|mths_since_last_major_derog|
+--------------------+-------------------+--------------------------+---------------------------+
|               504.0|              500.0|                       0.0|                       NULL|
|               589.0|              585.0|                       0.0|                       NULL|
|               574.0|              570.0|                       0.0|                       NULL|
|               614.0|              610.0|                       0.0|                       NULL|
|               534.0|              530.0|                       0.0|                       NULL|
|               514.0|              510.0|                       0.0|                       NULL|
|               554.0|              550.0|                       0.0|                       NULL|
|               499.

In [136]:
df.select("mths_since_last_major_derog").distinct().show(5) #need to dropoff

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 138, Finished, Available, Finished)

+---------------------------+
|mths_since_last_major_derog|
+---------------------------+
|                      170.0|
|                      147.0|
|                      169.0|
|                      160.0|
|                       70.0|
+---------------------------+
only showing top 5 rows



In [137]:
df=df.fillna(0,subset = ["last_fico_range_high","last_fico_range_low","collections_12_mths_ex_med"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 139, Finished, Available, Finished)

In [138]:
df.select(["policy_code","application_type","acc_now_delinq"]).show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 140, Finished, Available, Finished)

+-----------+----------------+--------------+
|policy_code|application_type|acc_now_delinq|
+-----------+----------------+--------------+
|        1.0|      Individual|           0.0|
|        1.0|      Individual|           0.0|
|        1.0|      Individual|           0.0|
|        1.0|      Individual|           0.0|
|        1.0|      Individual|           0.0|
|        1.0|      Individual|           0.0|
|        1.0|      Individual|           0.0|
|        1.0|      Individual|           0.0|
|        1.0|      Individual|           0.0|
|        1.0|      Individual|           0.0|
+-----------+----------------+--------------+
only showing top 10 rows



In [139]:
df.select("application_type").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 141, Finished, Available, Finished)

+----------------+
|application_type|
+----------------+
|       Joint App|
|      Individual|
+----------------+



In [140]:
df=df.fillna(0, subset = ["policy_code","acc_now_delinq"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 142, Finished, Available, Finished)

In [141]:
df.select(["tot_cur_bal","tot_cur_bal"]).show(5)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 143, Finished, Available, Finished)

+-----------+-----------+
|tot_cur_bal|tot_cur_bal|
+-----------+-----------+
|   194764.0|   194764.0|
|    86501.0|    86501.0|
|   233225.0|   233225.0|
|   612285.0|   612285.0|
|   401020.0|   401020.0|
+-----------+-----------+
only showing top 5 rows



In [142]:
df.select(["open_acc_6m","open_act_il","open_il_12m","open_il_24m"]).show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 144, Finished, Available, Finished)

+-----------+-----------+-----------+-----------+
|open_acc_6m|open_act_il|open_il_12m|open_il_24m|
+-----------+-----------+-----------+-----------+
|       NULL|       NULL|       NULL|       NULL|
|       NULL|       NULL|       NULL|       NULL|
|       NULL|       NULL|       NULL|       NULL|
|       NULL|       NULL|       NULL|       NULL|
|       NULL|       NULL|       NULL|       NULL|
|       NULL|       NULL|       NULL|       NULL|
|       NULL|       NULL|       NULL|       NULL|
|       NULL|       NULL|       NULL|       NULL|
|       NULL|       NULL|       NULL|       NULL|
|       NULL|       NULL|       NULL|       NULL|
+-----------+-----------+-----------+-----------+
only showing top 10 rows



In [143]:
df.select(["open_acc_6m","open_act_il","open_il_12m","open_il_24m"]).distinct().show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 145, Finished, Available, Finished)

+-----------+-----------+-----------+-----------+
|open_acc_6m|open_act_il|open_il_12m|open_il_24m|
+-----------+-----------+-----------+-----------+
|        4.0|        4.0|        4.0|        6.0|
|        2.0|        6.0|        1.0|        7.0|
|        2.0|        4.0|        4.0|        9.0|
|        3.0|        1.0|        4.0|        5.0|
|        3.0|       15.0|        0.0|        0.0|
|        2.0|        8.0|        2.0|        8.0|
|       10.0|        1.0|        1.0|        1.0|
|       11.0|        0.0|        0.0|        0.0|
|        3.0|        0.0|        1.0|        3.0|
|        0.0|       20.0|        3.0|        6.0|
+-----------+-----------+-----------+-----------+
only showing top 10 rows



In [144]:
df = df.fillna(0, subset = ["open_acc_6m","open_act_il","open_il_12m","open_il_24m"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 146, Finished, Available, Finished)

In [146]:
df.select(["mths_since_rcnt_il","total_bal_il","il_util","open_rv_12m","open_rv_24m","max_bal_bc"]).distinct().show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 148, Finished, Available, Finished)

+------------------+------------+-------+-----------+-----------+----------+
|mths_since_rcnt_il|total_bal_il|il_util|open_rv_12m|open_rv_24m|max_bal_bc|
+------------------+------------+-------+-----------+-----------+----------+
|              16.0|     29135.0|   77.0|        3.0|        4.0|    1326.0|
|              19.0|     67955.0|  102.0|        2.0|        4.0|    3283.0|
|               1.0|     52907.0|   83.0|        3.0|        6.0|    3876.0|
|               8.0|     15629.0|   92.0|        2.0|        3.0|    2695.0|
|               3.0|     50367.0|   70.0|        0.0|        1.0|   10953.0|
|               8.0|     17919.0|   45.0|        1.0|        5.0|    2474.0|
|             125.0|         0.0|   NULL|        2.0|        3.0|    7540.0|
|              89.0|         0.0|   NULL|        4.0|        7.0|    5793.0|
|              13.0|     18517.0|   73.0|        0.0|        2.0|    4010.0|
|              17.0|     11923.0|   52.0|        1.0|        3.0|    3198.0|

In [147]:
df=df.fillna(0, subset = ["mths_since_rcnt_il","total_bal_il","il_util","open_rv_12m","open_rv_24m","max_bal_bc"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 149, Finished, Available, Finished)

In [148]:
df.select(["all_util","total_rev_hi_lim","inq_fi","total_cu_tl","inq_last_12m","acc_open_past_24mths"]).distinct().show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 150, Finished, Available, Finished)

+--------+----------------+------+-----------+------------+--------------------+
|all_util|total_rev_hi_lim|inq_fi|total_cu_tl|inq_last_12m|acc_open_past_24mths|
+--------+----------------+------+-----------+------------+--------------------+
|    40.0|         47900.0|   1.0|        0.0|         3.0|                10.0|
|    51.0|         23700.0|   0.0|        0.0|         2.0|                 7.0|
|    NULL|        177900.0|  NULL|       NULL|        NULL|                 1.0|
|    NULL|         33700.0|  NULL|       NULL|        NULL|                 5.0|
|    NULL|         30000.0|  NULL|       NULL|        NULL|                 2.0|
|    NULL|         42500.0|  NULL|       NULL|        NULL|                 3.0|
|    NULL|         15100.0|  NULL|       NULL|        NULL|                 4.0|
|    NULL|          8500.0|  NULL|       NULL|        NULL|                 1.0|
|    NULL|         11300.0|  NULL|       NULL|        NULL|                 5.0|
|    NULL|         28750.0| 

In [149]:
df = df.fillna(0, subset = ["all_util","total_rev_hi_lim","inq_fi","total_cu_tl","inq_last_12m","acc_open_past_24mths"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 151, Finished, Available, Finished)

In [151]:
df.select(["avg_cur_bal","bc_open_to_buy","bc_util","chargeoff_within_12_mths","delinq_amnt"]).distinct().show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 153, Finished, Available, Finished)

+-----------+--------------+-------+------------------------+-----------+
|avg_cur_bal|bc_open_to_buy|bc_util|chargeoff_within_12_mths|delinq_amnt|
+-----------+--------------+-------+------------------------+-----------+
|    17629.0|        5742.0|   34.0|                     0.0|        0.0|
|     2806.0|        1943.0|   88.9|                     0.0|        0.0|
|     2268.0|         457.0|   86.9|                     0.0|        0.0|
|     4422.0|       32169.0|   35.9|                     0.0|        0.0|
|     8169.0|        4482.0|   75.8|                     0.0|        0.0|
|     5144.0|        4664.0|   86.6|                     0.0|        0.0|
|    30907.0|       11894.0|   63.8|                     0.0|        0.0|
|    13767.0|        4780.0|   75.5|                     0.0|        0.0|
|     3497.0|         482.0|   95.2|                     0.0|        0.0|
|     3746.0|        1528.0|   82.2|                     0.0|        0.0|
+-----------+--------------+-------+--

In [152]:
df =df.fillna(0,subset = ["avg_cur_bal","bc_open_to_buy","bc_util","chargeoff_within_12_mths","delinq_amnt"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 154, Finished, Available, Finished)

In [153]:
df.select(["mo_sin_old_il_acct","mo_sin_old_rev_tl_op","mo_sin_rcnt_rev_tl_op","mo_sin_rcnt_tl","mort_acc"]).distinct().show(10)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 155, Finished, Available, Finished)

+------------------+--------------------+---------------------+--------------+--------+
|mo_sin_old_il_acct|mo_sin_old_rev_tl_op|mo_sin_rcnt_rev_tl_op|mo_sin_rcnt_tl|mort_acc|
+------------------+--------------------+---------------------+--------------+--------+
|             121.0|               157.0|                  9.0|           3.0|     2.0|
|              44.0|               310.0|                 14.0|          11.0|     2.0|
|              45.0|                81.0|                  5.0|           5.0|     1.0|
|              80.0|               233.0|                141.0|          40.0|     3.0|
|             126.0|               107.0|                  8.0|           4.0|     0.0|
|              70.0|                82.0|                  8.0|           8.0|     0.0|
|              20.0|               205.0|                  0.0|           0.0|     0.0|
|              NULL|               194.0|                  4.0|           4.0|     0.0|
|             126.0|            

In [154]:
df=df.fillna(0, subset = ["mo_sin_old_il_acct","mo_sin_old_rev_tl_op","mo_sin_rcnt_rev_tl_op","mo_sin_rcnt_tl","mort_acc"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 156, Finished, Available, Finished)

In [156]:
df.select(["mths_since_recent_bc","mths_since_recent_inq","mths_since_recent_revol_delinq"]).show(5)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 158, Submitted, Running, Running)

+--------------------+---------------------+------------------------------+
|mths_since_recent_bc|mths_since_recent_inq|mths_since_recent_revol_delinq|
+--------------------+---------------------+------------------------------+
|                 8.0|                  7.0|                          NULL|
|                21.0|                  1.0|                          NULL|
|                 4.0|                 19.0|                          NULL|
|                 2.0|                  2.0|                          NULL|
|                11.0|                  8.0|                          NULL|
+--------------------+---------------------+------------------------------+
only showing top 5 rows



In [157]:
df=df.fillna(0,subset = ["mths_since_recent_bc","mths_since_recent_inq","mths_since_recent_revol_delinq"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 159, Finished, Available, Finished)

In [158]:
df.select(["num_accts_ever_120_pd","num_actv_bc_tl","num_actv_rev_tl","num_bc_sats","num_bc_tl","num_il_tl","num_op_rev_tl"]).distinct().show(5)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 160, Finished, Available, Finished)

+---------------------+--------------+---------------+-----------+---------+---------+-------------+
|num_accts_ever_120_pd|num_actv_bc_tl|num_actv_rev_tl|num_bc_sats|num_bc_tl|num_il_tl|num_op_rev_tl|
+---------------------+--------------+---------------+-----------+---------+---------+-------------+
|                  0.0|           8.0|           12.0|        9.0|      9.0|      0.0|         13.0|
|                  0.0|           1.0|            4.0|        3.0|      5.0|      5.0|          7.0|
|                  0.0|           2.0|            2.0|        4.0|      4.0|      1.0|          4.0|
|                  0.0|           2.0|            3.0|        3.0|      5.0|     11.0|          5.0|
|                  0.0|           2.0|            6.0|        3.0|      4.0|      4.0|          9.0|
+---------------------+--------------+---------------+-----------+---------+---------+-------------+
only showing top 5 rows



In [159]:
df = df.fillna(0, subset = ["num_accts_ever_120_pd","num_actv_bc_tl","num_actv_rev_tl","num_bc_sats","num_bc_tl","num_il_tl","num_op_rev_tl"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 161, Finished, Available, Finished)

In [160]:
df.select(["num_rev_accts","num_rev_tl_bal_gt_0","num_sats","num_tl_120dpd_2m","num_tl_30dpd","num_tl_90g_dpd_24m","num_tl_op_past_12m"]).distinct().show(5)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 162, Finished, Available, Finished)

+-------------+-------------------+--------+----------------+------------+------------------+------------------+
|num_rev_accts|num_rev_tl_bal_gt_0|num_sats|num_tl_120dpd_2m|num_tl_30dpd|num_tl_90g_dpd_24m|num_tl_op_past_12m|
+-------------+-------------------+--------+----------------+------------+------------------+------------------+
|         31.0|               16.0|    22.0|             0.0|         0.0|               0.0|               2.0|
|         30.0|                7.0|    18.0|             0.0|         0.0|               0.0|               2.0|
|          3.0|                3.0|    10.0|             0.0|         0.0|               0.0|               0.0|
|          8.0|                8.0|    10.0|             0.0|         0.0|               0.0|               1.0|
|         19.0|               11.0|    20.0|             0.0|         0.0|               0.0|               4.0|
+-------------+-------------------+--------+----------------+------------+------------------+---

In [161]:
df =df.fillna(0, subset = ["num_rev_accts","num_rev_tl_bal_gt_0","num_sats","num_tl_120dpd_2m","num_tl_30dpd","num_tl_90g_dpd_24m","num_tl_op_past_12m"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 163, Finished, Available, Finished)

In [162]:
df.select(["pct_tl_nvr_dlq","percent_bc_gt_75","pub_rec_bankruptcies","tax_liens","tot_hi_cred_lim","total_bal_ex_mort","total_bc_limit"]).distinct().show(5)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 164, Finished, Available, Finished)

+--------------+----------------+--------------------+---------+---------------+-----------------+--------------+
|pct_tl_nvr_dlq|percent_bc_gt_75|pub_rec_bankruptcies|tax_liens|tot_hi_cred_lim|total_bal_ex_mort|total_bc_limit|
+--------------+----------------+--------------------+---------+---------------+-----------------+--------------+
|         100.0|            62.5|                 0.0|      0.0|       346800.0|          79979.0|       67100.0|
|         100.0|             0.0|                 0.0|      0.0|         1300.0|              0.0|        1000.0|
|         100.0|            66.7|                 0.0|      0.0|        25000.0|          18085.0|       12300.0|
|         100.0|            66.7|                 0.0|      0.0|        48599.0|          40956.0|        6400.0|
|         100.0|            58.3|                 0.0|      0.0|       498462.0|          97740.0|       94400.0|
+--------------+----------------+--------------------+---------+---------------+--------

In [163]:
df=df.fillna(0, subset = ["pct_tl_nvr_dlq","percent_bc_gt_75","pub_rec_bankruptcies","tax_liens","tot_hi_cred_lim","total_bal_ex_mort","total_bc_limit"])

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 165, Finished, Available, Finished)

In [164]:
df.select(["total_il_high_credit_limit","hardship_flag","disbursement_method","debt_settlement_flag"]).distinct().show(5)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 166, Finished, Available, Finished)

+--------------------------+-------------+-------------------+--------------------+
|total_il_high_credit_limit|hardship_flag|disbursement_method|debt_settlement_flag|
+--------------------------+-------------+-------------------+--------------------+
|                  123139.0|            N|               Cash|                   N|
|                   36944.0|            N|               Cash|                   N|
|                   21000.0|            N|          DirectPay|                   N|
|                   94043.0|            N|               Cash|                   Y|
|                   43789.0|            N|               Cash|                   Y|
+--------------------------+-------------+-------------------+--------------------+
only showing top 5 rows



In [165]:
df=df.fillna(0, subset = "total_il_high_credit_limit")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 167, Finished, Available, Finished)

In [166]:
df.select("hardship_flag").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 168, Finished, Available, Finished)

+-------------+
|hardship_flag|
+-------------+
|            Y|
|            N|
+-------------+



In [167]:
df.select("disbursement_method").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 169, Finished, Available, Finished)

+-------------------+
|disbursement_method|
+-------------------+
|               Cash|
|          DirectPay|
+-------------------+



In [168]:
df.select("debt_settlement_flag").distinct().show()

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 170, Finished, Available, Finished)

+--------------------+
|debt_settlement_flag|
+--------------------+
|                   Y|
|                   N|
+--------------------+



In [169]:
col_to_drop = ['member_id', 'desc', 'mths_since_last_record', 'annual_inc_joint', 'dti_joint', 'verification_status_joint', 'mths_since_recent_bc_dlq', 'revol_bal_joint', 'sec_app_fico_range_low', 'sec_app_fico_range_high', 'sec_app_earliest_cr_line', 'sec_app_inq_last_6mths', 'sec_app_mort_acc', 'sec_app_open_acc', 'sec_app_revol_util', 'sec_app_open_act_il', 'sec_app_num_rev_accts', 'sec_app_chargeoff_within_12_mths', 'sec_app_collections_12_mths_ex_med', 'sec_app_mths_since_last_major_derog', 'hardship_type', 'hardship_reason', 'hardship_status', 'deferral_term', 'hardship_amount', 'hardship_start_date', 'hardship_end_date', 'payment_plan_start_date', 'hardship_length', 'hardship_dpd', 'hardship_loan_status', 'orig_projected_additional_accrued_interest', 'hardship_payoff_balance_amount', 'hardship_last_payment_amount', 'debt_settlement_flag_date', 'settlement_status', 'settlement_date', 'settlement_amount', 'settlement_percentage', 'settlement_term','Term','emp_length','url','mths_since_last_major_derog']

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 171, Finished, Available, Finished)

In [173]:
col_to_drop #44

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 175, Finished, Available, Finished)

['member_id',
 'desc',
 'mths_since_last_record',
 'annual_inc_joint',
 'dti_joint',
 'verification_status_joint',
 'mths_since_recent_bc_dlq',
 'revol_bal_joint',
 'sec_app_fico_range_low',
 'sec_app_fico_range_high',
 'sec_app_earliest_cr_line',
 'sec_app_inq_last_6mths',
 'sec_app_mort_acc',
 'sec_app_open_acc',
 'sec_app_revol_util',
 'sec_app_open_act_il',
 'sec_app_num_rev_accts',
 'sec_app_chargeoff_within_12_mths',
 'sec_app_collections_12_mths_ex_med',
 'sec_app_mths_since_last_major_derog',
 'hardship_type',
 'hardship_reason',
 'hardship_status',
 'deferral_term',
 'hardship_amount',
 'hardship_start_date',
 'hardship_end_date',
 'payment_plan_start_date',
 'hardship_length',
 'hardship_dpd',
 'hardship_loan_status',
 'orig_projected_additional_accrued_interest',
 'hardship_payoff_balance_amount',
 'hardship_last_payment_amount',
 'debt_settlement_flag_date',
 'settlement_status',
 'settlement_date',
 'settlement_amount',
 'settlement_percentage',
 'settlement_term',
 'Term'

In [175]:
len(df.columns) #totalcomuns

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 177, Finished, Available, Finished)

153

In [176]:
df=df.drop(*col_to_drop) #dropping the un necessary columns

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 178, Finished, Available, Finished)

In [177]:
len(df.columns)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 179, Finished, Available, Finished)

109

In [178]:
display(df.head(5))

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 180, Finished, Available, Finished)

SynapseWidget(Synapse.DataFrame, 54c912f3-509d-4e1c-919e-04edc387a8ad)

In [180]:
df.select("id").distinct().show(5)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 182, Finished, Available, Finished)

+--------+
|      id|
+--------+
|56622220|
|12597709|
|48324574|
|48110888|
|59103178|
+--------+
only showing top 5 rows



In [181]:
df.groupBy(col("id")).count().show(30)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 183, Finished, Available, Finished)

+--------+-----+
|      id|count|
+--------+-----+
|56622220|    1|
|12597709|    1|
|48324574|    1|
|48110888|    1|
|59103178|    1|
|57971109|    1|
|55152456|    1|
|57005442|    1|
|53464632|    1|
|34592983|    1|
|58150691|    1|
|58370305|    1|
|58693161|    1|
|52818986|    1|
|55991703|    1|
|47551221|    1|
|52748295|    1|
|34412729|    1|
|11408061|    1|
|33391643|    1|
|11655937|    1|
|57883113|    1|
|58060167|    1|
|57335371|    1|
|56021594|    1|
|58524152|    1|
|10735779|    1|
|10638934|    1|
|35753400|    1|
|36890207|    1|
+--------+-----+
only showing top 30 rows



In [182]:
df.count(),len(df.columns)

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 184, Finished, Available, Finished)

(2260668, 109)

In [196]:
df.write.format("delta").saveAsTable("Cleaned_bank_data")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 198, Finished, Available, Finished)

In [195]:
df.write.option("header",True).csv("Files/Cleaned_bank_data.csv")

StatementMeta(, 3cca4f1d-ae58-44c8-ae2a-a16209ab7dbc, 197, Finished, Available, Finished)

# Finally Banking Dataset cleaning completed. 
# Rawdatset : Total colmuns -  151, Total Rows - 22,60,701
# Cleaned dataset : Total colmuns -  109, Total Rows - 22,60,668
# Newly added columns count - 3
# columns removed due to nulls more than 75% - 41
